Improve handling of pgsql-stored layer styles and field widgets

strk opened the pull request:
See #26186

strk committed:
Use CREATE IF NOT EXISTS for layer_styles

The syntax is supported since PostgreSQL 9.1 and we already use
9.1-only syntax (regclass).

strk committed:
Use EXISTS to tell if a table exists, rather than count(*)

Avoids a full sequencial scan (see #26186)

Also query pg_class directly

strk committed:
Reduce number of queries to qgis_editor_widget_styles table

Closes #26186

strk commented:
I left it to have a Travis backed proof of the removed code NOT being needed for what the test was testing. Just in case my test run was bogus...

So I take you agree with the removal @pvalsecc  ?

pvalsecc commented:
> So I take you agree with the removal @pvalsecc ?

No, not at all.
But I don't have time to fix it... and I don't have the energy to fight for it.
So suit yourself.

strk commented:
I did the fixing work already so it's not much energy to keep it, but it takes energy to PROVE it is needed. If the only existing test passes even w/out that code, what's the point of keeping it ?

It doesn't take a fight to keep the code, just a document describing how it works and a test proving it works.

I'm happy to give advice a to how to make the support not bring down performances

haubourg commented:
> It doesn't take a fight to keep the code, just a document describing how it works and a test proving it works.

Much agreed, this is the second time this hidden feature takes a lot of time to a some of us. And causes performance issues for users. 

@pvalsecc  Can you please (again) document the feature in the code and push a link to the eventual tool using it (hopefully it is a public plugin). If this feature is for a non publishable use and it is not possible to document it, I tend to think we should remove it faster than proposed in the dev's list, ie mark it deprecated and wait for 4.0 to remove it.  cc @nyalldawson @yjacolin

strk commented:
@haubourg can you confirm the code in this PR fixed the speed issue for you ?

strk commented:
Travis fail with a segfault while the relevant test on my machine works just fine, I'm confused

m-kuhn commented:
I still tend to think we shouldn't just drop it on such a short notice.

What about this:

 - Add a QSettings based opt-in approach to still enable this feature with 3.10.
 - Deprecate it for removal with 3.12 (the next non-LTR)

Eventually this might be used by a production system somewhere. Dropping support now for 3.10 is too quick to react for any business process and potentially leaves customers with mixed feelings about the reliability of QGIS.

@pvalsecc that would give you a year before it's dropped from LTR (and 4 months before it's dropped from the code). Is this enough time to check with your project lead / the client how we can come to a future proof solution?

strk commented:
I like the QSettings opt-in approach idea. It would also give us a chance to specify exact schema in which the table would be found, further improving the code reading it..

strk commented:
As I keep getting Travis failures of various kind (last one was an indentation error when the change reduced to ONLY removing lines) I've now force-pushed a no-change commit (all commits reverted). Will slowly re-introduce commits if there's a green, but reading actual errors from Travis is really revealing problematic.

m-kuhn commented:
Did you setup the pre commit hook? That should sort out all the indentation and formatting.

strk commented:
Yes my pre-commit hook is in place, as a symlink to ../../scripts/prepare-commit.sh (and I do see it working)

strk commented:
Ok, I now understand that the existing test will actually FAIL if we remove the support (Travis confirms this). What the test does is it checks that `editorWidgetSetup()` called against a QgsField of a QgsVectorLayer loaded from PostgreSQL  is not null, when a configuration is specified, for that field, in the `qgis_editor_widget_styles` table.

pvalsecc commented:
Happy to here I didn't mess up too much back then :-)

strk commented:
I get a segfault from PyQgsPostgresProvider test, upon calling `conn->PQexec( sql );`
The sql string seems fine, but that call never returns, so I suspect it's a deleted connection, not sure how it gets deleted...

strk commented:
It seems to be the destruction of a QgsPostgresResult:
```
#0  0x00007f1eb778f7d5 in PQclear () at /usr/lib/x86_64-linux-gnu/libpq.so.5
#1  0x00007f1e84467215 in QgsPostgresResult::~QgsPostgresResult() (this=0x7ffd4d4d49b0, __in_chrg=<optimized out>)
    at /usr/src/qgis/qgis-improve-metadata-tables-creation-and-usage/src/providers/postgres/qgspostgresconn.cpp:52
#2  0x00007f1e844245a1 in QgsPostgresProviderMetadata::saveStyle(QString const&, QString const&, QString const&, QString const&, QString const&, QString const&, bool, QString&) (this=
    0x34769e0, uri=..., qmlStyle=..., sldStyle=..., styleName=..., styleDescription=..., uiFileContent=..., useAsDefault=false, errCause=...)
    at /usr/src/qgis/qgis-improve-metadata-tables-creation-and-usage/src/providers/postgres/qgspostgresprovider.cpp:4757
```

I guess I made the scope of those variables too wide

strk commented:
Interesting (or not?) re-using a QgsPostgresResult seems to trigger the segfault. Using another object doesn't have problems

strk commented:
I think I found the culprit.
This segfaults:
> res = QgsPostgresResult( conn->PQexec( checkQuery ) );

This is fine:
> res = conn->PQexec( checkQuery );

@jef-n I suspect we'd want to add some protection in the copy costructor for QgsPostgresResult...

strk commented:
@m-kuhn @pvalsecc  are you ok with merging this PR as-is (no QSetting, just improved query, passing existing test) ? We can do the QSettings in 3.10

nyalldawson commented:
Hey @strk I'm lost with the status on this one - my understanding is that the current pr is just optimisations which doesn't change current behaviour. Is that correct?

strk commented:
@nyalldawson that's correct, no behavioral change, just optimization

strk commented:
@jgrocha I saw you changed the code for layer_styles recently, can you please check if things still work for you with this PR ?

m-kuhn commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -1126,36 +1126,58 @@ bool QgsPostgresProvider::loadFields()
+    }
 
+    // Set corresponding editor widget
+    for ( auto &field : mAttributeFields )
```
```suggestion
    for ( auto &field : qgis::as_const( mAttributeFields ) )
```

strk replied:
field needs be non-const in order to be able to do field.setEditorWidgetSetup

m-kuhn replied:
Most classes will detach and work on a copy of the container if iterating them non const.
Does this code here work as expected with QgsFields?

strk replied:
As tests do exist and used to fail when I made a mistake I'd say it's working as-is

m-kuhn commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -1126,36 +1126,58 @@ bool QgsPostgresProvider::loadFields()
+  }
+
+  QStringList quotedFnames;
+  foreach ( const QString &name, mAttributeFields.names() )
```
```suggestion
  const QStringList fieldNames;
  for ( const QString &name : fieldNames )
```

strk replied:
this suggestion isn't clear, partial (fieldNames will be empty?)

m-kuhn replied:
Sorry, updated.

pvalsecc commented:
> @m-kuhn @pvalsecc are you ok with merging this PR as-is

Yes, I'm OK with the PR. Thanks a lot.

strk committed:
Use for instead of foreach

m-kuhn commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -1132,7 +1132,7 @@ void QgsPostgresProvider::setEditorWidgets()
 
   QStringList quotedFnames;
-  foreach ( const QString &name, mAttributeFields.names() )
+  for ( const QString &name : mAttributeFields.names() )
```
This will detach

strk replied:
Where can I read more about this "detach" thing ?
I'm iterating it as const (you mentioned a condition being iterating as non-const?).
Also, why   const QStringList fieldNames = mAttributeFields.names() instead of holding a const reference  ? Let's take the chance to see if test will fail with current code ?

strk replied:
btw, I'm sure there's a more elegant way to call a function for each element in a container...

m-kuhn replied:
Tests won't fail. It's just a matter of loosing some memory and cpu cycles.
There's a thorough explanation here: https://www.kdab.com/goodbye-q_foreach/

m-kuhn replied:
This one looks nice https://stackoverflow.com/questions/10516495/calling-a-function-on-every-element-of-a-c-vector I've never tested it though

m-kuhn reviewed (changes requested):
I had a good reason to suggest the changes the way I did ;)

strk committed:
Iterate over a const, to supposedly save some CPU cycles and memory

... as suggested by @m-khun via
https://github.com/qgis/QGIS/pull/31858#discussion_r339471008

m-kuhn reviewed (approved):
Thanks, I'll ignore the doubtful commit message

jgrocha reviewed (approved):
I'm ok with the PR. Thanks for improving the initial PR.

strk commented:
Would this be good to backport to 3.10 @m-kuhn @nyalldawson ?

m-kuhn commented:
Yes, I think so. Should we wait for a couple of days in master before merging the backport or is it low-risk?

strk commented:
Given how young 3.10 is I'd say it's low risk. We'll get a bunch of issues reported for 3.10 anyway, so we'll be around to fix them

m-kuhn commented:
The risk isn't related to the age of 3.10 ;)

strk commented:
Backport PR: #32573
